File: src/core/LogViewMachine.ts
export interface LogEntry {
  id: string;
  timestamp: Date;
  level: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';
  message: string;
  data?: any;
  tags?: string[];
}

export interface LogFilter {
  level?: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';
  tags?: string[];
  startTime?: Date;
  endTime?: Date;
  limit?: number;
}

export interface LogViewMachineConfig {
  maxEntries?: number;
  defaultFilter?: LogFilter;
  enableRealTime?: boolean;
}

export interface LogViewMachine {
  getLogs(filter?: LogFilter): LogEntry[];
  addLog(entry: Omit<LogEntry, 'id' | 'timestamp'>): void;
  clearLogs(): void;
  subscribe(callback: (logs: LogEntry[]) => void): () => void;
}

export class LogViewMachineImpl implements LogViewMachine {
  private logs: LogEntry[] = [];
  private subscribers: Set<(logs: LogEntry[]) => void> = new Set();
  private config: LogViewMachineConfig;

  constructor(config: LogViewMachineConfig = {}) {
    this.config = {
      maxEntries: 1000,
      enableRealTime: true,
      ...config
    };
  }

  getLogs(filter?: LogFilter): LogEntry[] {
    let filteredLogs = [...this.logs];

    if (filter) {
      if (filter.level) {
        const levelOrder = { 'DEBUG': 0, 'INFO': 1, 'WARN': 2, 'ERROR': 3 };
        const minLevel = levelOrder[filter.level];
        filteredLogs = filteredLogs.filter(log => levelOrder[log.level] >= minLevel);
      }

      if (filter.tags && filter.tags.length > 0) {
        filteredLogs = filteredLogs.filter(log => 
          log.tags && filter.tags!.some(tag => log.tags!.includes(tag))
        );
      }

      if (filter.startTime) {
        filteredLogs = filteredLogs.filter(log => log.timestamp >= filter.startTime!);
      }

      if (filter.endTime) {
        filteredLogs = filteredLogs.filter(log => log.timestamp <= filter.endTime!);
      }

      if (filter.limit) {
        filteredLogs = filteredLogs.slice(-filter.limit);
      }
    }

    return filteredLogs;
  }

  addLog(entry: Omit<LogEntry, 'id' | 'timestamp'>): void {
    const newEntry: LogEntry = {
      ...entry,
      id: Math.random().toString(36).substr(2, 9),
      timestamp: new Date()
    };

    this.logs.push(newEntry);

    if (this.config.maxEntries && this.logs.length > this.config.maxEntries) {
      this.logs = this.logs.slice(-this.config.maxEntries);
    }

    if (this.config.enableRealTime) {
      this.notifySubscribers();
    }
  }

  clearLogs(): void {
    this.logs = [];
    this.notifySubscribers();
  }

  subscribe(callback: (logs: LogEntry[]) => void): () => void {
    this.subscribers.add(callback);
    callback(this.logs); // Initial call

    return () => {
      this.subscribers.delete(callback);
    };
  }

  private notifySubscribers(): void {
    this.subscribers.forEach(callback => callback(this.logs));
  }
} 

File: log-view-machine/core/src/main/kotlin/com/logview/core/LogViewMachine.kt
package com.logview.core

import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

interface LogViewMachine {
    fun getLogs(filter: LogFilter? = null): List<LogEntry>
    fun addLog(entry: LogEntry)
    fun clearLogs()
    fun subscribe(callback: (List<LogEntry>) -> Unit): () -> Unit
}

data class LogEntry(
    val id: String,
    val timestamp: Instant,
    val level: String,
    val message: String,
    val metadata: Map<String, Any>? = null,
    val tags: List<String>? = null
)

data class LogFilter(
    val level: String? = null,
    val tags: List<String>? = null,
    val startTime: Instant? = null,
    val endTime: Instant? = null,
    val limit: Int? = null
)

class LogViewMachineImpl(
    private val config: LogViewMachineConfig = LogViewMachineConfig()
) : LogViewMachine {
    private val logs = mutableListOf<LogEntry>()
    private val subscribers = mutableSetOf<(List<LogEntry>) -> Unit>()

    data class LogViewMachineConfig(
        val maxEntries: Int = 1000,
        val enableRealTime: Boolean = true
    )

    override fun getLogs(filter: LogFilter?): List<LogEntry> {
        var filteredLogs = logs.toList()

        if (filter != null) {
            if (filter.level != null) {
                val levelOrder = mapOf("DEBUG" to 0, "INFO" to 1, "WARN" to 2, "ERROR" to 3)
                val minLevel = levelOrder[filter.level] ?: 0
                filteredLogs = filteredLogs.filter { log ->
                    val logLevel = levelOrder[log.level] ?: 0
                    logLevel >= minLevel
                }
            }

            if (filter.tags != null && filter.tags.isNotEmpty()) {
                filteredLogs = filteredLogs.filter { log ->
                    log.tags?.any { tag -> filter.tags.contains(tag) } == true
                }
            }

            if (filter.startTime != null) {
                filteredLogs = filteredLogs.filter { log ->
                    log.timestamp >= filter.startTime
                }
            }

            if (filter.endTime != null) {
                filteredLogs = filteredLogs.filter { log ->
                    log.timestamp <= filter.endTime
                }
            }

            if (filter.limit != null) {
                filteredLogs = filteredLogs.takeLast(filter.limit)
            }
        }

        return filteredLogs
    }

    override fun addLog(entry: LogEntry) {
        val newEntry = entry.copy(
            id = if (entry.id.isEmpty()) System.currentTimeMillis().toString() else entry.id,
            timestamp = if (entry.timestamp == Instant.EPOCH) Instant.now() else entry.timestamp
        )

        logs.add(newEntry)

        // Maintain max entries limit
        if (config.maxEntries > 0 && logs.size > config.maxEntries) {
            logs.removeAt(0)
        }

        if (config.enableRealTime) {
            notifySubscribers()
        }
    }

    override fun clearLogs() {
        logs.clear()
        notifySubscribers()
    }

    override fun subscribe(callback: (List<LogEntry>) -> Unit): () -> Unit {
        subscribers.add(callback)
        callback(logs) // Initial call
        return { subscribers.remove(callback) }
    }

    private fun notifySubscribers() {
        subscribers.forEach { callback ->
            try {
                callback(logs)
            } catch (e: Exception) {
                // Log error but don't break other subscribers
                println("Error in log subscriber: ${e.message}")
            }
        }
    }
} 